File: ios/Internal/AuthenticationManager.swift
import Foundation
import LocalAuthentication
import Security

/// Protocol for managing authentication and biometric operations.
///
/// Encapsulates all authentication prompt handling, LAContext creation,
/// and biometric evaluation. Follows Single Responsibility Principle.
///
/// @since 6.0.0
protocol AuthenticationManager {
  /// Create LAContext configured for the given authentication prompt.
  ///
  /// - Parameters:
  ///   - prompt: Optional authentication prompt with customization
  /// - Returns: Configured LAContext for biometric evaluation
  func makeLAContext(prompt: AuthenticationPrompt?) -> LAContext

  /// Perform Keychain query with authentication if needed.
  ///
  /// - Parameters:
  ///   - query: Base Keychain query
  ///   - prompt: Optional authentication prompt
  /// - Returns: Query result or nil if not found
  /// - Throws: RuntimeError if authentication fails
  func executeAuthenticatedQuery(
    _ query: [String: Any],
    prompt: AuthenticationPrompt?
  ) throws -> AnyObject?

  /// Check if error indicates authentication was canceled.
  ///
  /// - Parameters:
  ///   - status: OSStatus from Keychain operation
  /// - Returns: True if user canceled authentication
  func isAuthenticationCanceled(status: OSStatus) -> Bool

  /// Create error message for authentication-related OSStatus.
  ///
  /// - Parameters:
  ///   - status: OSStatus from operation
  /// - Returns: RuntimeError with appropriate message
  func makeAuthenticationError(for status: OSStatus) -> RuntimeError
}

File: android/src/main/java/com/sensitiveinfo/internal/auth/AuthenticationManager.kt
package com.sensitiveinfo.internal.auth

import com.margelo.nitro.sensitiveinfo.AuthenticationPrompt

/**
 * Interface for managing authentication and biometric operations on Android.
 *
 * Encapsulates all biometric authentication, device credential handling,
 * and authentication prompt management. Follows Single Responsibility Principle.
 *
 * @since 6.0.0
 */
interface AuthenticationManager {
  /**
   * Check if biometric authentication is available.
   *
   * @return True if device supports biometric authentication
   */
  suspend fun isBiometricAvailable(): Boolean

  /**
   * Check if device credential (PIN/Pattern/Password) is available.
   *
   * @return True if device has set device credentials
   */
  suspend fun isDeviceCredentialAvailable(): Boolean

  /**
   * Evaluate biometric authentication with prompt.
   *
   * @param prompt Optional authentication prompt with customization
   * @return True if authentication succeeded
   * @throws Exception if authentication fails or is canceled
   */
  suspend fun evaluateBiometric(prompt: AuthenticationPrompt?): Boolean

  /**
   * Evaluate device credential authentication with prompt.
   *
   * @param prompt Optional authentication prompt with customization
   * @return True if authentication succeeded
   * @throws Exception if authentication fails or is canceled
   */
  suspend fun evaluateDeviceCredential(prompt: AuthenticationPrompt?): Boolean

  /**
   * Check if error indicates authentication was canceled by user.
   *
   * @param exception The exception from authentication
   * @return True if user canceled
   */
  fun isAuthenticationCanceled(exception: Exception): Boolean

  /**
   * Create error message for authentication-related errors.
   *
   * @param exception The authentication exception
   * @return Formatted error message
   */
  fun makeAuthenticationError(exception: Exception): String
}
